internal/uvm/loopback.go

```diff
@@ -0,0 +1,159 @@
+package uvm
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Microsoft/hcsshim/internal/guestrequest"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/requesttype"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const lcowLoopbackLayerFmt = "/run/layers/lb%d"
+
+// LoopbackDevice represents a loopback device in the guest.
+type LoopbackDevice struct {
+	// Utility VM the loopback device
+	vm *UtilityVM
+	// deviceNumber is the device number for the loopback device
+	deviceNumber uint32
+	// refCount for this loopback device if the same
+	refCount uint32
+	// backingFile is the backing file for the loopback device. This is the path to
+	// a file in the guest and not the host.
+	backingFile string
+	// path in the UVM the device is mounted at.
+	uvmPath string
+}
+
+// Release removes the loopback device from the UVM it belongs to.
+func (lb *LoopbackDevice) Release(ctx context.Context) error {
+	if err := lb.vm.RemoveLoopback(ctx, lb.backingFile); err != nil {
+		return errors.Wrapf(err, "failed to remove loopback device with backing file %s", lb.backingFile)
+	}
+	return nil
+}
+
+// findNextLoopback finds the next available loopback device.
+//
+// The lock MUST be held when calling this function.
+func (uvm *UtilityVM) findNextLoopbackDevice(ctx context.Context) uint32 {
+	for k, v := range uvm.loopbackDevices {
+		if v == nil {
+			log.G(ctx).WithFields(logrus.Fields{
+				"deviceNumber": k,
+			}).Debug("found previous loopback device free")
+			return k
+		}
+	}
```

Comment:
I don't think we really care about reusing indices here do we?

Reply:
We do, so the less loop devices we actually have to allocate the better here. We start off with 8 available (/dev/loop0.../dev/loop7) and after that we'll have to start making new ones. When we actually "remove" the device from the guest all we're going to do is remove the backing file so the node is free to use again. 

Could imagine this situation:

Start 8 layer container  -> /dev/loop0 through /dev/loop7 in use -> Stop container
Start 8 layer container -> /dev/loop0 through /dev/loop7 free again -> Re-use -> Stop container

Ideally we should just be able to re-use the first 8 (or any we allocate also as long as they're not in use) without having to allocate any further